lib/pbench/test/unit/server/test_file_tree.py

```diff
@@ -83,15 +79,93 @@ def test_clean_emptys(self, server_config, make_logger):
             for r in roots:
                 assert not (r / c).exists()
 
-    def test_lifecycle(self, monkeypatch, server_config, make_logger, tarball):
+    def test_create_bad(self, monkeypatch, server_config, make_logger, tarball):
+        # Calling restorecon() gives warning messages about "no default label"
+        monkeypatch.setattr("selinux.restorecon", lambda a: None)
+
+        source_tarball, source_md5, md5 = tarball
+        dataset_name = Tarball.stem(source_tarball)
+        tree = FileTree(server_config, make_logger)
+
+        # Attempting to create a dataset from the md5 file should result in
+        # a bad filename error
+        with pytest.raises(BadFilename) as exc:
+            tree.create("ABC", source_md5)
+        assert type(exc.value) is BadFilename
+        assert exc.value.path == str(source_md5)
+
+        tree.create("ABC", source_tarball)
 
+        # The create will remove the source files, so trying again should
+        # result in an error.
+        with pytest.raises(BadFilename) as exc:
+            tree.create("ABC", source_md5)
+        assert type(exc.value) is BadFilename
+        assert exc.value.path == str(source_md5)
+
+        # Attempting to create the same dataset again (from the archive copy)
+        # should fail with a duplicate dataset error.
+        tarball = tree.find_dataset(dataset_name)
+        with pytest.raises(DuplicateDataset) as exc:
+            tree.create("ABC", tarball.tarball_path)
+        assert type(exc.value) is DuplicateDataset
+        assert str(exc.value) == "A dataset named 'pbench-user-benchmark_some + config_2021.05.01T12.42.42' is already present in the file tree"
+        assert exc.value.dataset == tarball.name
+
+    def test_find(self, monkeypatch, server_config, make_logger, tarball):
         # Calling restorecon() gives warning messages about "no default label"
         monkeypatch.setattr("selinux.restorecon", lambda a: None)
 
         source_tarball, source_md5, md5 = tarball
+        dataset_name = Tarball.stem(source_tarball)
         tree = FileTree(server_config, make_logger)
         tree.create("ABC", source_tarball)
 
+        # The original files should have been removed
+        assert not source_tarball.exists()
+        assert not source_md5.exists()
+
+        tarball = tree.find_dataset(dataset_name)
+        assert tarball
+        assert tarball.name == dataset_name
+
+        # Test __getitem__
+        assert tarball == tree[dataset_name]
+        with pytest.raises(DatasetNotFound) as exc:
+            tree["foobar"]
+        assert type(exc.value) is DatasetNotFound
+        assert str(exc.value) == "The dataset named 'foobar' is not present in the file tree"
```

Comment:
Apparently, [`pytest.raises()`](https://docs.pytest.org/en/6.2.x/reference.html#pytest-raises) will let you specify the message text as part of the expected-match, so you don't have to assert it separately (although, perhaps doing it separately produces "better" code).

Reply:
I concur....